src/spatial_data_format.rs

```diff
@@ -0,0 +1,559 @@
+            .collect();
+        Self::attach_tags(&self.header.tags, quantized_streams)
+    }
+}
+
+/// This builder contains the data required
+#[derive(Debug, Clone)]
+pub struct GrapeFileBuilder {
+    sample_rate: u64,
+    streams: Vec<(GrapeTag, Vec<f32>)>,
+}
+
+impl Default for GrapeFileBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GrapeFileBuilder {
+    /// Instaintiate a builder with no streams and a default sample rate of
+    /// 1000 samples per second.
+    fn new() -> Self {
+        GrapeFileBuilder {
+            sample_rate: 1000,
+            streams: Vec::new(),
+        }
+    }
+
+    /// Sets the sample rate of the builder to the argument.
+    pub fn set_samplerate(self, sample_rate: u64) -> Self {
+        GrapeFileBuilder {
+            sample_rate,
+            ..self
+        }
+    }
+
+    /// Adds a tagged stream to the builder.
+    pub fn add_stream(mut self, stream: &[f32], tag: GrapeTag) -> Self {
+        let stream: Vec<f32> = stream.to_vec();
+        self.streams.push((tag, stream));
+        self
+    }
+
+    /// Removes all streams from the builder
+    pub fn clear_streams(mut self) -> Self {
+        self.streams.clear();
+        self
+    }
+
+    /// Builds a [GrapeFile] from the builder, truncating all streams to the
+    /// length of the shortest stream.
+    pub fn build_truncate(self) -> GrapeFile {
+        let tags: Vec<GrapeTag> = self
+            .streams
+            .iter()
+            .map(|(tag, _vec)| tag)
+            .cloned()
+            .collect();
+        let sample_vecs: Vec<Vec<f32>> = self.streams.into_iter().map(|(_tag, vec)| vec).collect();
+
+        let mut samples = Vec::new();
+
+        if !sample_vecs.is_empty() {
+            let shortest = sample_vecs.iter().map(|v| v.len()).min().unwrap();
+            samples.reserve_exact(shortest * sample_vecs.len() as usize);
+            for sample_idx in 0..shortest {
+                for stream_idx in 0..sample_vecs.len() {
+                    samples.push(sample_vecs[stream_idx as usize][sample_idx]);
+                }
+            }
+        };
+
+        GrapeFile {
+            header: GrapeFileHeader {
+                n_streams: sample_vecs.len() as u64,
+                sample_rate: self.sample_rate,
+                tags,
+            },
+            samples,
+        }
+    }
+
+    /// Builds a [GrapeFile] from the builder, extending all streams with the
+    /// last value recorded in each stream.
+    pub fn build_extend(self) -> GrapeFile {
+        let tags: Vec<GrapeTag> = self
+            .streams
+            .iter()
+            .map(|(tag, _vec)| tag)
+            .cloned()
+            .collect();
+        let sample_vecs: Vec<Vec<f32>> = self.streams.into_iter().map(|(_tag, vec)| vec).collect();
+
+        let mut samples = Vec::new();
+
+        if !sample_vecs.is_empty() {
+            let longest = sample_vecs.iter().map(|v| v.len()).max().unwrap();
```

Comment:
Could we avoid this unwrap by doing:
```suggestion
        let longest = sample_vecs.iter().map(|v| v.len()).max();
        if let Some(longest) = longest {
```
I know it's equivalent but it's nice to avoid unwraps :)

Reply:
Agh yes. I neglected to clear those up before opening the PR. Maybe we should put clippy in the CI :)